In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [3]:
# dataset: https://www.kaggle.com/ronitf/heart-disease-uci

In [4]:
heart_data = pd.read_csv("/content/heart.csv")

In [5]:
heart_data.head()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

In [6]:
heart_data.tail()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
298   57    0   0       140   241    0  ...      1      0.2      1   0     3       0
299   45    1   3       110   264    0  ...      0      1.2      1   0     3       0
300   68    1   0       144   193    1  ...      0      3.4      1   2     3       0
301   57    1   0       130   131    0  ...      1      1.2      1   1     3       0
302   57    0   1       130   236    0  ...      0      0.0      1   1     2       0

[5 rows x 14 columns]

In [9]:
heart_data.shape

(303, 14)

In [7]:
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [8]:
heart_data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [11]:
heart_data.describe()

age         sex          cp  ...          ca        thal      target
count  303.000000  303.000000  303.000000  ...  303.000000  303.000000  303.000000
mean    54.366337    0.683168    0.966997  ...    0.729373    2.313531    0.544554
std      9.082101    0.466011    1.032052  ...    1.022606    0.612277    0.498835
min     29.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000
25%     47.500000    0.000000    0.000000  ...    0.000000    2.000000    0.000000
50%     55.000000    1.000000    1.000000  ...    0.000000    2.000000    1.000000
75%     61.000000    1.000000    2.000000  ...    1.000000    3.000000    1.000000
max     77.000000    1.000000    3.000000  ...    4.000000    3.000000    1.000000

[8 rows x 14 columns]

In [13]:
# distribution of target variable

heart_data['target'].value_counts()

1    165
0    138
Name: target, dtype: int64

165 have heart disease,
138 have no heart disease

# Split the independent and dependent variables

In [14]:
# As our distribution is almsot balanced we do not undersample


X = heart_data.drop(columns='target', axis = 1)
Y = heart_data['target']

In [15]:
print(X)

     age  sex  cp  trestbps  chol  ...  exang  oldpeak  slope  ca  thal
0     63    1   3       145   233  ...      0      2.3      0   0     1
1     37    1   2       130   250  ...      0      3.5      0   0     2
2     41    0   1       130   204  ...      0      1.4      2   0     2
3     56    1   1       120   236  ...      0      0.8      2   0     2
4     57    0   0       120   354  ...      1      0.6      2   0     2
..   ...  ...  ..       ...   ...  ...    ...      ...    ...  ..   ...
298   57    0   0       140   241  ...      1      0.2      1   0     3
299   45    1   3       110   264  ...      0      1.2      1   0     3
300   68    1   0       144   193  ...      0      3.4      1   2     3
301   57    1   0       130   131  ...      1      1.2      1   1     3
302   57    0   1       130   236  ...      0      0.0      1   1     2

[303 rows x 13 columns]


In [16]:
print(Y)

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64


# Splitting the data into Train and Test data

In [17]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=123)

stratify selects with the same distribution

In [18]:
print(X.shape, X_train.shape, X_test.shape)

(303, 13) (242, 13) (61, 13)


# Model Training

In [19]:
# logistic regression

model = LogisticRegression()

In [20]:
# train the log res model with train data

model.fit(X_train, Y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

As our dataset is small, it is better to have more observations

# Model Evaluation

In [22]:
# accuracy on train data

X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [23]:
print("Accuracy on Train data: ", training_data_accuracy)

Accuracy on Train data:  0.8553719008264463


In [25]:
# accuracy on test data

X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [26]:
print("Accuracy on Test data: ", test_data_accuracy)

Accuracy on Test data:  0.8360655737704918


# Building a Predictive System

input the results of medical anaysis (all columns except the last one)  of a person and the system will predict if the person has heart disease.

In [114]:
input_data = input("""Please write the results of medical analysis with the following order: \n 
age \n 
sex (0 for man, 1 for woman) \n
chest pain type (4 values)  \n
resting blood pressure  \n
serum cholestoral in mg/dl  \n 
fasting blood sugar > 120 mg/dl \n 
resting electrocardiographic results (values 0,1,2) \n 
maximum heart rate achieved \n 
exercise induced angina \n 
oldpeak = ST depression induced by exercise relative to rest \n 
the slope of the peak exercise ST segment \n 
number of major vessels (0-3) colored by flourosopy \n 
thal: 3 = normal; 6 = fixed defect; 7 = reversable defect \n 
please use single space after each value""")

# change the input data to a numpy array

new_input_data = [int(s) for s in input_data.split() if s.isdigit()]

input_data_as_numpy_array = np.asarray(new_input_data)

# we used reshape the numpy array because we predict only one person
input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)

prediction = model.predict(input_data_reshaped)

print(prediction)

if (prediction[0] == 0):
  print("\n\n\n THE PERSON DOES NOT HAVE HEART DISEASE""")

else:
  print("\n\n\n THE PERSON HAS HEART DISEASE")

Please write the results of medical analysis with the following order: 
 
age 
 
sex (0 for man, 1 for woman) 

chest pain type (4 values)  

resting blood pressure  

serum cholestoral in mg/dl  
 
fasting blood sugar > 120 mg/dl 
 
resting electrocardiographic results (values 0,1,2) 
 
maximum heart rate achieved 
 
exercise induced angina 
 
oldpeak = ST depression induced by exercise relative to rest 
 
the slope of the peak exercise ST segment 
 
number of major vessels (0-3) colored by flourosopy 
 
thal: 3 = normal; 6 = fixed defect; 7 = reversable defect 
 
please use single space after each value39 0 0 165 250 0 0 155 0 4 0 2 2
[0]



 THE PERSON DOES NOT HAVE HEART DISEASE
